File: python/day09.py
import bisect


def get_first_non_sum(nums, k):
    window = sorted(nums[:k])
    for i in range(k, len(nums)):
        l, r = 0, k - 1
        s = nums[i]
        while l < r:
            cur_sum = window[l] + window[r]
            if cur_sum < s:
                l += 1
            elif cur_sum > s:
                r -= 1
            else:
                break
        if l == r:
            return s
        bisect.insort(window, s)
        del window[bisect.bisect_left(window, nums[i - k])]


def find_sum_range(nums, s):
    l, r = 0, 0
    cur_sum = nums[0]
    n = len(nums)
    while r < n - 1 and l < n - 1:
        if cur_sum < s:
            r += 1
            cur_sum += nums[r]
        elif cur_sum > s:
            cur_sum -= nums[l]
            l += 1
        else:
            break
    if l != r:
        return (l, r)


def solution():
    nums = [int(g) for g in open('../data/09.txt', 'r').readlines()]

    WINDOW_SIZE = 25

    non_sum = get_first_non_sum(nums, WINDOW_SIZE)
    print(f'Answer 1: {non_sum}')

    (b, e) = find_sum_range(nums, non_sum)
    print(f'Answer 2: {min(nums[b:e + 1]) + max(nums[b:e + 1])}')

File: rust/src/day09.rs
use crate::common;
use std::collections::BTreeMap;

fn get_first_non_sum(nums: &Vec<i64>, k: usize) -> Option<i64> {
    let mut window = BTreeMap::new();
    let n = nums.len();
    for i in 0..k {
        *window.entry(nums[i]).or_insert(1) += 1;
    }
    for i in k..n {
        let s = nums[i];
        let (mut it_l, mut it_r) = (window.iter(), window.iter().rev());
        let (mut l, mut r) = (it_l.next().unwrap(), it_r.next().unwrap());
        while l.0 != r.0 {
            let cur_sum = l.0 + r.0;
            if cur_sum < s {
                l = it_l.next().unwrap();
            } else if cur_sum > s {
                r = it_r.next().unwrap();
            } else {
                break;
            }
        }
        if l.0 == r.0 && *l.1 == 1 {
            return Some(s);
        }
        let to_remove = nums[i - k];
        window.entry(to_remove).and_modify(|e| *e -= 1);
        if window[&to_remove] == 0 {
            window.remove(&to_remove);
        }
        *window.entry(s).or_insert(1) += 1;
    }
    None
}

fn find_sum_range(nums: &Vec<i64>, s: i64) -> Option<(usize, usize)> {
    let (mut l, mut r) = (0, 0);
    let mut cur_sum = nums[0];
    let n = nums.len();
    while r < n - 1 && l < n - 1 {
        if cur_sum < s {
            r += 1;
            cur_sum += nums[r];
        } else if cur_sum > s {
            cur_sum -= nums[l];
            l += 1;
        } else {
            break;
        }
    }
    if l != r {
        Some((l, r))
    } else {
        None
    }
}

fn find_sum_range_agg(nums: &Vec<i64>, s: i64) -> Option<i64> {
    let sum_range = find_sum_range(&nums, s)?;
    let range_it = &nums[sum_range.0..=sum_range.1];
    let range_max = range_it.iter().max()?;
    let range_min = range_it.iter().min()?;
    Some(range_min + range_max)
}

pub(crate) fn solution() {
    const WINDOW_SIZE: usize = 25;
    let nums: Vec<i64> = common::read_integers(&common::data_file(9)).unwrap();
    let non_sum = get_first_non_sum(&nums, WINDOW_SIZE).unwrap();
    println!("First non-sum: {}", non_sum);
    println!("Sum range: {}", find_sum_range_agg(&nums, non_sum).unwrap());
}
